File: Solutions/72. Edit Distance/72. Edit Distance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int sz1=word1.size(),sz2=word2.size();
        vector<vector<int>> dp(2,vector<int>(sz1+1));
        for(int i=0;i<=sz1;++i)
            dp[0][i]=i;
        for(int i=1;i<=sz2;++i)
            for(int j=0;j<=sz1;++j){
                if(j==0)
                    dp[i%2][j]=i;
                else if(word1[j-1]==word2[i-1])
                    dp[i%2][j]=dp[(i-1)%2][j-1];
                else
                    dp[i%2][j]=1+min(dp[(i-1)%2][j-1]
                                        ,min(dp[(i-1)%2][j],dp[(i%2)][j-1]));
            }
        return dp[sz2%2][sz1];
    }
};

File: Solutions/72. Edit Distance/72. Edit Distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        sz1,sz2=len(word1),len(word2)
        # dp[1] Nested lists =>wrong
        # dp1=[[0]*(sz1+1)]*2
        
        # two-dimensional array => right
        dp=[[0]*(sz1+1) for _ in range(2)]
        #print(dp1)
        print(dp)
        for i in range(sz1+1):
            dp[0][i]=i
        for i in range(1,sz2+1):
            for j in range(sz1+1):
                if j==0:
                    dp[i%2][j]=i
                elif word1[j-1]==word2[i-1]:
                    dp[i%2][j]=dp[(i-1)%2][j-1]
                else:
                    dp[i%2][j]=1+min(dp[(i-1)%2][j-1],dp[(i-1)%2][j],dp[(i%2)][j-1])
        return dp[sz2%2][sz1]